### PR TITLE
AVRO-3475: Enforce  time-millis and time-micros specification

### PR DESCRIPTION
**Pull request by zcsizmadia**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO-3475)
  - https://issues.apache.org/jira/browse/AVRO-3475
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

**Comment by zcsizmadia** [comment]
@anekdoti This is the PR we discussed in #1623

File: lang/csharp/src/apache/main/Util/TimeMicrosecond.cs [hunk 1]
```diff
@@ -51,16 +51,29 @@ public override object ConvertToBaseValue(object logicalValue, LogicalSchema sch
...
+
+            // Note: UnixEpochDateTime.TimeOfDay is '00:00:00', so the Add is meaningless. This could be 'return time;'
+            return UnixEpochDateTime.TimeOfDay.Add(time);
+        }
+
+        private static void ThrowIfOutOfRange(TimeSpan time, string paramName)
+        {
+            if (time.Ticks < 0 || time >= _exclusiveUpperBound)
+            {
+                throw new ArgumentOutOfRangeException(paramName, $"A '{LogicalTypeName}' value must be at least '00:00:00' and less than '1.00:00:00'.");
```
Comment (KyleSchoonover) [inline lang/csharp/src/apache/main/Util/TimeMicrosecond.cs:75]: `throw new ArgumentOutOfRangeException(paramName, $"A '{LogicalTypeName}' value must be at least '{TimeSpan.Zero}' and less than '{_exclusiveUpperBound}'.");`

Use the constants when possible for messaging.

File: lang/csharp/src/apache/main/Util/TimeMillisecond.cs [hunk 2]
```diff
@@ -50,17 +50,29 @@ public override object ConvertToBaseValue(object logicalValue, LogicalSchema sch
...
+
+            // Note: UnixEpochDateTime.TimeOfDay is '00:00:00'. This could be 'return time;'
+            return UnixEpochDateTime.TimeOfDay.Add(time);
+        }
+
+        private static void ThrowIfOutOfRange(TimeSpan time, string paramName)
+        {
+            if (time.Ticks < 0 || time >= _exclusiveUpperBound)
+            {
+                throw new ArgumentOutOfRangeException(paramName, $"A '{LogicalTypeName}' value must be at least '00:00:00' and less than '1.00:00:00'.");
```
Comment (KyleSchoonover) [inline lang/csharp/src/apache/main/Util/TimeMillisecond.cs:74]: `throw new ArgumentOutOfRangeException(paramName, $"A '{LogicalTypeName}' value must be at least '{TimeSpan.Zero}' and less than '{_exclusiveUpperBound}'.");`

Use the constants when possible for messaging.

**Review by KyleSchoonover — APPROVED** [review]

**Comment by martin-g** [comment]
Thank you, @zcsizmadia !

**Pull request merged**